---
title: 'ConsensusMetaDA: an R Package for consensus-based differential abundance analysis of metagenomic data.'
author: "Manoharan Kumar, Matt Field"
date: 'Last modified: 2025-07-23. Compiled: `r Sys.Date()`'
output:
  html_document:
    highlight: tango
    number_sections: yes
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: no
      smooth_scroll: no
  pdf_document:
    toc: yes
    toc_depth: '3'
  word_document:
    toc: yes
    toc_depth: '3'
vignette: >
  %\VignetteIndexEntry{ConsensusMetaDA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction to ConsensusMetaDA

ConsensusMetaDA is an R package for microbiome analysis using multiple algorithms - reaching consensus. No one tool perform best DA anlayes on metaganome data. Hence it needs a consensus approach to obtain more robust differential abundant microbiome. ConsensusMetaDA helps achieve this using five popularly used tools. ConsensusMetaDA uses popular biom format and samples table as input to perform DA along with it generates standard microbiome visualization plots such as Alpha Diversity, Beta Diversity, Rarefaction curve, Bidirectional plot and Scale plot at various taxa levels.

ConsensusMetaDA is simplified into two functions:

* ```build_OTU_counts()``` generate an phyloseq object that contain OTU reads counts (using biome, sample counts and taxa files)
* ```OTU_multi_DA()``` perform DA analysis (all possible pairwise comparisons)

Below the core functionality of ConsensusMetaDA as well as how to plot results using the OTU_plots function.


```{r echo=TRUE, message=FALSE, warning=FALSE}
# install supporting packages

# ADAPT R package requires R version 4.4 hence need all R updated to R v4.4.2

# One line per package
# if (!require("ggplot2", quietly = TRUE)) install.packages("ggplot2")
# if (!require("BiocManager", quietly = TRUE)) install.packages("BiocManager")
# if (!require("phyloseq", quietly = TRUE)) BiocManager::install("phyloseq")
# if (!require("DESeq2", quietly = TRUE)) BiocManager::install("DESeq2")
# if (!require("ALDEx2", quietly = TRUE)) BiocManager::install("ALDEx2")
# if (!require("edgeR", quietly = TRUE)) BiocManager::install("edgeR")
# if (!require("ADAPT", quietly = TRUE)) BiocManager::install("ADAPT")
# if (!require("metagenomeSeq", quietly = TRUE)) BiocManager::install("metagenomeSeq")
# 
# if (!require("ConsensusMetaDA", quietly = TRUE)) {
#   if (!require("remotes", quietly = TRUE)) install.packages("remotes")
#   remotes::install_github("kmanoharan01/ConsensusMetaDA")
# }

```


## Load Libraries
Begin by first installing and loading the ConsensusMetaDA library.

```{r echo=TRUE, message=FALSE, warning=FALSE}

# Load the packages
library(ggplot2)
library(phyloseq)
library(DESeq2)
library(ALDEx2)
library(edgeR)
library(metagenomeSeq)
library(ADAPT)


# load ConsensusMetaDA

library(ConsensusMetaDA)

```

# ConsensusMetaDA examples

To illustrate functionality of ```ConsensusMetaDA```, we will utilise microbiome biom format data from the ```ecc_saliva``` and annotation libraries as follows. Begin by installing and attaching data from these libraries as follows:



```{r echo=TRUE, message=FALSE, warning=FALSE}
library(ADAPT)

data("ecc_saliva")

otu_mat <- as(otu_table(ecc_saliva), "matrix")
sample_df <- as(sample_data(ecc_saliva), "data.frame")
tax_mat <- as(tax_table(ecc_saliva), "matrix")

#otu_final <- data.frame(sampleID = rownames(otu_mat), otu_mat)

#write.table(otu_mat, file = "ecc_saliva_Otu.txt",  col.names = TRUE, sep = "\t", quote = FALSE)

#write.table(sample_df, file = "ecc_saliva_sample_table.txt", col.names = TRUE, sep = "\t", quote = FALSE)

#write.table(tax_mat, file = "ecc_saliva_tax_table.txt", col.names = TRUE, sep = "\t", quote = FALSE)

################

#system("module load python3/3.10.4")

#system("~/.local/bin/biom convert -i ecc_saliva_Otu.txt   -o ecc_saliva_Otu.biom --table-type='OTU table' --to-json")

ecc_saliva_biom <- "../inst//extdata/ecc_saliva_Otu.biom"

ecc_saliva_sample_table <- "../inst//extdata/ecc_saliva_sample_table.txt"

ecc_saliva_tax <- tax_mat

taxa_names(tax_mat)
#tax_mat_test <- tax_table(tax_mat)

```

# Building a phyloseq format 

This function takes inputs such as biom file, sample list, and taxa information and creates a build_OTU_count object. To build a ConsensusMetaDA requires a biom file and sample grouping table with location path. Taxa level by accepts 7 taxa level. This function is incorporated utilizing the phyloseq (v1.50.0) package (McMurdie & Holmes, 2013).  It also can take filtering parameters such as abundance_threshold, prevalence_threshold, rarity_threshold and variance_threshold

This generates phyloseq object that stores all relevant information for performing differential abundance analysis. build_OTU_counts() allows users to build a phyloseq object by simply providing 1) a biom file, 2) a sample file, 3) a taxa and 4) filtering option. Below we will use biom file and sample table file as an example for creating a summarized experiment:


```{r echo=TRUE, message=FALSE, warning=FALSE}
## build
ecc_saliva_build_OTU_counts <- build_OTU_counts(ecc_saliva_biom, 
                                                 ecc_saliva_sample_table, 
                                                 ecc_saliva_tax, 
                                                 abundance_threshold = 10, 
                                                 prevalence_threshold = 0.1, 
                                                 rarity_threshold = 0, 
                                                 variance_threshold = 0 
)

ecc_saliva_build_OTU_counts


```

## filtering rarity format

With rarity filter one can provide numeric values and filter based on this value.

```{r echo=TRUE, message=FALSE, warning=FALSE}

ecc_saliva_build_OTU_counts <- build_OTU_counts(ecc_saliva_biom, 
                                                 ecc_saliva_sample_table, 
                                                 ecc_saliva_tax, 
                                                 abundance_threshold = 0, 
                                                 prevalence_threshold = 0, 
                                                 rarity_threshold = 0.1, 
                                                 variance_threshold = 0 
)

ecc_saliva_build_OTU_counts

```

## filtering abundance format

With abundance filter one can provide numeric values and filter based on this value.

```{r echo=TRUE, message=FALSE, warning=FALSE}

ecc_saliva_build_OTU_counts <- build_OTU_counts(ecc_saliva_biom, 
                                                 ecc_saliva_sample_table, 
                                                 ecc_saliva_tax, 
                                                 abundance_threshold = 10, 
                                                 prevalence_threshold = 0, 
                                                 rarity_threshold = 0, 
                                                 variance_threshold = 0 
)

ecc_saliva_build_OTU_counts

```

## filtering prevalence format

With prevalence filter one can provide numeric values and filter based on this value.


```{r echo=TRUE, message=FALSE, warning=FALSE}

ecc_saliva_build_OTU_counts <- build_OTU_counts(ecc_saliva_biom, 
                                                 ecc_saliva_sample_table, 
                                                 ecc_saliva_tax, 
                                                 abundance_threshold = 10, 
                                                 prevalence_threshold = 0.1, 
                                                 rarity_threshold = 0, 
                                                 variance_threshold = 0 
)

ecc_saliva_build_OTU_counts

```

## filtering variance format

With variance filter one can provide numeric values and filter based on this value.

```{r echo=TRUE, message=FALSE, warning=FALSE}

ecc_saliva_build_OTU_counts <- build_OTU_counts(ecc_saliva_biom, 
                                                 ecc_saliva_sample_table, 
                                                 ecc_saliva_tax, 
                                                 abundance_threshold = 0, 
                                                 prevalence_threshold = 0, 
                                                 rarity_threshold = 0, 
                                                 variance_threshold = 0.2 
)

ecc_saliva_build_OTU_counts

```

# Performing differential abundance

Performs DA analysis on all possible pairs of “groups” on the provided sample table using all available methods (currently edgeR, DESeq2, metagenomeSeq, ADAPT and ALDEx2). Also, OTUs_multi_DE produces a Venn diagram and upset plots for all comparisons. We have also implemented rank-sum (Guo, Zhao, Ye, Sheng, & Shyr, 2014) to perform p-value-based ranking in the summary table. The function creates a table for each comparison. The summary table contains statics including “p_union”, “p_intersect” and "rank_sum". The summary table provides results from three tools in one format hence easy to compare. “AveExpre” is the mean value of normalized average counts.

```{r echo=TRUE, message=FALSE, warning=FALSE}
## build
ecc_saliva_build_OTU_counts <- build_OTU_counts(ecc_saliva_biom, 
                                                 ecc_saliva_sample_table, 
                                                 ecc_saliva_tax, 
                                                 abundance_threshold = 10, 
                                                 prevalence_threshold = 0.1, 
                                                 rarity_threshold = 0, 
                                                 variance_threshold = 0 
)

ecc_saliva_build_OTU_counts

## Perform DA

ecc_saliva_DA <- OTUs_multi_DA(ecc_saliva_build_OTU_counts)


```

Running ```OTUs_multi_DA()``` will perform DA analysis on all possible pairs of "groups" and save these results as a simple list of "merged" - being the merged results of "DESeq2", "metagenomeSeq", "edgeR", "ALDEx2" and "ADAPT" into one table, as well as the latter five as objects independently. The data frame is sorted by the ```rank_sum```. The following columns are included:

* ```ID``` - Identifier
* ```LogFC``` - Log2 Fold-Change, also known as a log-ratio (average of edgeR, DESeq2, ALDEx2, MetagenomeSeq and ADAPT)
* ```LogFC_sd``` - Log2 Fold-Change standard deviation of LogFC (average) 
* ```edger_adj_p``` - EdgeR p-value adjusted for multiple hypotheses
* ```deseq_adj_p``` - DESeq2 p-value adjusted for multiple hypotheses
* ```ALDEx2_adj_p``` - ALDEx2 p-value adjusted for multiple hypotheses
* ```MetagenomeSeq_adj_p``` - MetagenomeSeq p-value adjusted for multiple hypotheses
* ```ADAPT_adj_p``` - ADAPT p-value adjusted for multiple hypotheses
* ```rank_sum``` - sum of the ranks from edgeR_rank, DESeq2_rank, edgeR, DESeq2, ALDEx2, MetagenomeSeq, ADAPT and rank_sum
* ```p_intersect``` - the largest p-value observed from all methods tested. // Empirical Benjamini-Hochberg (BH) adjusted p-value for the between group.
    * This represents the intersect when a threshold is set on the p_intersect column
* ```p_union``` - the smallest p-value observed from all methods tested. // Empirical Benjamini-Hochberg (BH) adjusted p-value for the between group.
    * This represents the union when a threshold is set on the p_union column To access the merged results:    

## UpSet Plot from Differential abundance

UpSet plot is plotted for each differential abundance comparison respectively.

```{r echo=TRUE, message=FALSE, warning=FALSE}

print(ecc_saliva_DA$Control_vs_Case$upset_plot)


```

## Venn Diagram  from Differential abundance

Venn Diagram is plotted for each differential abundance comparison respectively.

```{r echo=TRUE, message=FALSE, warning=FALSE}
library(grid)
grid.newpage()
grid.draw(ecc_saliva_DA$Control_vs_Case$VennDiagram)

```

# Plotting functions

Creates Five different plots and saves them as pdf files. (1) Alpha diversity - Shannon, (2) Beta diversity – Principal Coordinate Analyses (PcoA), (3) rare faction curve and (4) Scaled plot at taxa level 5) Bi-directional plots. These plots are generated using the ggplot2 package (Wickham, 2016).

In addition, each of the 5 plots can be plotted individually using the ```OTU_plots``` function. See ```?OTU_plots``` for description, which provides wrappers for 5 different plots. Next we will plot each of these using the example data.

```{r echo=TRUE, message=FALSE, warning=FALSE}

ecc_saliva_plots <- OTUs_plots(ecc_saliva_build_OTU_counts)

```

## rarefaction curve

A rarefaction curve is a simple graph that shows how many different microbial species you find as you sequence more and more DNA from a sample.

* Different sequencing depths shows unfair comparison
* Rarefaction normalizes all samples to same depth
* Now one can compare diversity between samples

Rarefaction curves help ensure one have sequenced enough to capture the true diversity in your samples and allow fair comparisons between different samples or studies.

```{r echo=TRUE, message=FALSE, warning=FALSE}

ecc_saliva_plots <- OTUs_plots(ecc_saliva_build_OTU_counts)

```

## Alpha Diversity
Alpha Diversity is present using Phyloseq Shannon index method; It represents within one sample microbiome diversity (e.g., one person's gut microbiome). Measures both richness and evenness of microbial species. Single-sample metric (unlike beta diversity which compares between samples). Higher values generally indicate healthier, more stable microbiomes.


```{r echo=TRUE, message=FALSE}

print(ecc_saliva_plots$AlphaDiv$tax_level)

```

## Beta Diversity/Principle Coordinates Analysis
Beta diversity or Principle Coordinates analysis represents difference among groups and diversity of microbiome and among samples. It's essential for understanding how microbiomes vary across individuals, conditions, or treatments.


```{r echo=TRUE, message=FALSE}

print(ecc_saliva_plots$BetaDiv)

```


## Bidirectional Plot

Bidirectional Plot shows two groups mean abundance and prevalance.
A bidirectional plot showing both abundance and prevalence displays two different types of microbial changes simultaneously on the same graph. This provides much richer information than either measurements alone. 

```{r echo=TRUE, message=FALSE, warning=FALSE}

print(ecc_saliva_plots$all_comparisons_results$Control_vs_Case$Genus_plot)

```

This bidirectional abundance-prevalence plot provides a powerful tool for understanding the complex dynamics of microbial communities, revealing patterns that would be missed by examining either abundance or prevalence alone. Here are descriptions :

* High abundance and High prevalence increase
  * Beneficial bacteria expanding after prebiotic treatment

* Higher abundance and Lower prevalence
	* Pathogen blooming in susceptible individuals only

* Lower abundance and Higher prevalence
	* Opportunistic bacteria colonizing stressed communities

* Lower abundance and Lower prevalence
	* Sensitive bacteria dying off after antibiotic treatment




# Citing results that use ```ConsesnsusMetaDA```

When using this package, please cite ConsesnsusMetaDA as follows and all methods used in your analysis.

For consensus DE:
```{r echo=TRUE, message=FALSE, warning=FALSE}
#citation("ConsesnsusMetaDA")
```

* When using DESeq2 (also check package reference suggestions)
    * Love, M.I., Huber, W., Anders, S. Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2 Genome Biology 15(12):550 (2014)

* When using edgeR (also check package reference suggestions)
    * Robinson MD, McCarthy DJ and Smyth GK (2010). edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics 26, 139-140
    * McCarthy DJ, Chen Y and Smyth GK (2012). Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation. Nucleic Acids Research 40, 4288-4297

* When using ALDEx2 (also check package reference suggestions)
	* Fernandes, A. D., Macklaim, J. M., Linn, T. G., Reid, G., & Gloor, G. B. (2013). ANOVA-Like Differential Expression (ALDEx) Analysis for Mixed Population RNA-Seq. PLOS ONE, 8(7), e67019. doi:10.1371/journal.pone.0067019
	
* When using ADAPT (also check package reference suggestions)
	* Wang, M., Fontaine, S., Jiang, H., & Li, G. (2024). ADAPT: Analysis of Microbiome Differential Abundance by Pooling Tobit Models. Bioinformatics, 40(11). doi:10.1093/bioinformatics/btae661

* When using MetagenomeSeq (also check package reference suggestions)
	* Paulson JN, O. N., Braccia DJ, Wagner J, Talukder H, Pop M, Bravo HC (2013). metagenomeSeq: Statistical analysis for sparse high-throughput sequncing. Bioconductor package.  Retrieved from http://www.cbcb.umd.edu/software/metagenomeSeq.
